src/modules/flow/mqtt/mqtt.c

```diff
@@ -0,0 +1,452 @@
+{
+    struct client_data *mdata = data;
+
+    if (mdata->mqtt)
+        sol_mqtt_disconnect(mdata->mqtt);
+
+    if (mdata->payload)
+        sol_blob_unref(mdata->payload);
+
+    sol_cert_free(mdata->ca_cert);
+    sol_cert_free(mdata->client_cert);
+    sol_cert_free(mdata->ca_cert);
+
+    free(mdata->host);
+    free(mdata->user);
+    free(mdata->pass);
+    free(mdata->id);
+    free(mdata->topic);
+}
+
+static int
+mqtt_client_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct client_data *mdata = data;
+    const struct sol_flow_node_type_mqtt_client_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options, SOL_FLOW_NODE_TYPE_MQTT_CLIENT_OPTIONS_API_VERSION, -EINVAL);
+    opts = (const struct sol_flow_node_type_mqtt_client_options *)options;
+
+    mdata->node = node;
+
+#define ALLOC_AND_CHECK(src, dst, alloc) \
+    do { \
+        if (src) { \
+            dst = alloc(src); \
+            SOL_NULL_CHECK_GOTO(dst, error); \
+        } \
+    } while (0) \
+
+    ALLOC_AND_CHECK(opts->host, mdata->host, strdup);
+    ALLOC_AND_CHECK(opts->username, mdata->user, strdup);
+    ALLOC_AND_CHECK(opts->password, mdata->pass, strdup);
+    ALLOC_AND_CHECK(opts->client_id, mdata->id, strdup);
+    ALLOC_AND_CHECK(opts->ca_cert, mdata->ca_cert, sol_cert_load_from_file);
+    ALLOC_AND_CHECK(opts->client_cert, mdata->client_cert, sol_cert_load_from_file);
+    ALLOC_AND_CHECK(opts->private_key, mdata->private_key, sol_cert_load_from_file);
+    ALLOC_AND_CHECK(opts->topic, mdata->topic, strdup);
+
+#undef ALLOC_AND_CHECK
+
+    mdata->port = opts->port;
+    mdata->keepalive = opts->keepalive;
+    mdata->qos = (sol_mqtt_qos)opts->qos;
+    mdata->clean_session = opts->clean_session;
+
+    return 0;
+
+error:
+    mqtt_client_close(node, data);
+    return -ENOMEM;
+}
+
+static int
+clean_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct client_data *mdata = data;
+    int r;
+    bool in_value;
+
+    r = sol_flow_packet_get_boolean(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->clean_session = in_value;
+
+    return 0;
+}
+
+static int
+keepalive_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct client_data *mdata = data;
+    int r;
+    struct sol_irange in_value;
+
+    r = sol_flow_packet_get_irange(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->keepalive = in_value.val;
+
+    return 0;
+}
+
+static int
+host_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct client_data *mdata = data;
+    int r;
+    const char *in_value;
+
+    r = sol_flow_packet_get_string(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    free(mdata->host);
+    mdata->host = strdup(in_value);
```

Comment:
Hmm... this actually uncovered a problem. sol_util_replace_str_if_changed() should be applicable here only if mdata->host is always valid, but in some scenarios, like in Bluemix example, the host is unknown when the flow start.

What would be better: let host be compulsory and keep it as is, which will make the user put a dummy value in the node config, or let if be optional and default to something like 127.0.0.1?